plutus-scb/src/Control/Monad/Freer/WebSocket.hs

```diff
@@ -0,0 +1,41 @@
+{-# LANGUAGE DataKinds        #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs            #-}
+{-# LANGUAGE TemplateHaskell  #-}
+{-# LANGUAGE TypeOperators    #-}
+
+module Control.Monad.Freer.WebSocket where
+
+import           Control.Monad.Freer           (Eff, LastMember, type (~>), interpret)
+import           Control.Monad.Freer.TH        (makeEffect)
+import           Control.Monad.IO.Class        (MonadIO, liftIO)
+import           Data.Aeson                    (FromJSON, ToJSON)
+import qualified Data.Aeson                    as JSON
+import           Data.UUID                     (UUID)
+import           Data.UUID.V4                  (nextRandom)
+import qualified Network.WebSockets            as WS
+import           Network.WebSockets.Connection (Connection, PendingConnection, receiveData)
+
+data WebSocketEffect r where
```

Comment:
You are writing an effect that is supposed to generically model how a web-socket would work, but in its syntax definition, you are using a type from the library used to write the interpreter. 

This means that you are coupling the syntax to a specific interpreter, in which case you cannot easily use a different way of interpreting this effect.

I am not sure that this is wrong, as I am not well-versed in web-sockets.

Reply:
I don't think that's really the case here. `WebSocketEffect` defines the model. It's the interpreter `handleWebSocket` that ties it to a particular library. But that's okay - the handler is exactly where we want to tie the model to a specific implementation.

I can easily foresee adding a different handler called `handleMockWebsocket` or something, that would just give dummy responses. The interface wouldn't change.

That to me is the important part - the caller gets an abstract interface that doesn't depend on the implementation.

Reply:
Maybe @raduom means the use of the types `Connection` and `PendingConnection`. Do these types have exposed constructors?

Reply:
it seems that it is, so I think that's fine personally

Reply:
@shmish111 that is what I meant. But as both you and @krisajenkins noticed, it's not at all problematic.

Reply:
Yeah, I'd missed your point @raduom. Sorry! Thanks for spelling it out for me @shmish111. 😊